File: kotlin/src/test/kotlin/com/cyaim/im/client/GapRepairTest.kt
@file:OptIn(ExperimentalCoroutinesApi::class)

package com.cyaim.im.client

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlin.coroutines.EmptyCoroutineContext
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * The correctness property this SDK exists for: the application never sees a conversation jump
 * forward and then fill in behind it.
 *
 * The cold-start half of the same story — what happens across a *process* restart rather than a
 * socket one — lives in [CursorTest], because it is a different failure with a different fix.
 */
class GapRepairTest {

    @Test
    fun `a gap is repaired before the message that revealed it is delivered`() = runTest {
        val gateway = gatewayWith(
            syncReply = { request -> syncPage(request, "c1", listOf(2L, 3L, 4L)) },
        )
        val client = newClient(gateway)
        val received = collectMessages(client)
        openClient(client, gateway)

        gateway.latest.push(PushTarget.Message, messageJson("c1", 1))
        runCurrent()

        // seq 5 with a local cursor of 1: 2, 3 and 4 were never delivered.
        gateway.latest.push(PushTarget.Message, messageJson("c1", 5))
        runCurrent()

        assertEquals(listOf(1L, 2L, 3L, 4L, 5L), received.map { it.seq })

        val sync = assertNotNull(gateway.latest.requestsTo("msg.sync").singleOrNull())
        assertEquals(2L, sync.bodyLong("fromSeq"))
        assertEquals(4L, sync.bodyLong("toSeq"))
        assertEquals(3L, sync.bodyLong("limit"))
    }

    @Test
    fun `a contiguous stream never calls msg dot sync`() = runTest {
        val gateway = gatewayWith()
        val client = newClient(gateway)
        val received = collectMessages(client)
        openClient(client, gateway)

        for (seq in 1L..5L) {
            gateway.latest.push(PushTarget.Message, messageJson("c1", seq))
            runCurrent()
        }

        assertEquals(listOf(1L, 2L, 3L, 4L, 5L), received.map { it.seq })
        assertTrue(gateway.latest.requestsTo("msg.sync").isEmpty())
    }

    @Test
    fun `a replayed message after a reconnect is dropped silently`() = runTest {
        val gateway = gatewayWith()
        val client = newClient(gateway)
        val received = collectMessages(client)
        openClient(client, gateway)

        gateway.latest.push(PushTarget.Message, messageJson("c1", 1))
        runCurrent()
        gateway.latest.push(PushTarget.Message, messageJson("c1", 2))
        runCurrent()
        gateway.latest.push(PushTarget.Message, messageJson("c1", 2))
        runCurrent()

        assertEquals(listOf(1L, 2L), received.map { it.seq })
    }

    @Test
    fun `an online-only message never moves the cursor`() = runTest {
        val gateway = gatewayWith(syncReply = { request -> syncPage(request, "c1", listOf(2L)) })
        val client = newClient(gateway)
        val received = collectMessages(client)
        openClient(client, gateway)

        gateway.latest.push(PushTarget.Message, messageJson("c1", 1))
        runCurrent()

        // seq 0: a chatroom or online-only message that was never persisted. It is delivered, but
        // it must not be mistaken for progress — otherwise the next real message looks contiguous.
        gateway.latest.push(PushTarget.Message, messageJson("c1", 0))
        runCurrent()
        gateway.latest.push(PushTarget.Message, messageJson("c1", 3))
        runCurrent()

        assertEquals(listOf(1L, 0L, 2L, 3L), received.map { it.seq })
    }

    @Test
    fun `gaps in different conversations do not block each other`() = runTest {
        val gateway = gatewayWith(
            syncReply = { request ->
                syncPage(request, request.bodyString("conversationId"), listOf(2L))
            },
        )
        val client = newClient(gateway)
        val received = collectMessages(client)
        openClient(client, gateway)

        gateway.latest.push(PushTarget.Message, messageJson("c1", 1))
        gateway.latest.push(PushTarget.Message, messageJson("c2", 1))
        runCurrent()

        gateway.latest.push(PushTarget.Message, messageJson("c1", 3))
        gateway.latest.push(PushTarget.Message, messageJson("c2", 3))
        runCurrent()

        assertEquals(2, gateway.latest.requestsTo("msg.sync").size)
        assertEquals(listOf(1L, 2L, 3L), received.filter { it.conversationId == "c1" }.map { it.seq })
        assertEquals(listOf(1L, 2L, 3L), received.filter { it.conversationId == "c2" }.map { it.seq })
    }

    /**
     * A device that has been offline for a week is behind by more messages than it can usefully
     * receive at once. The cursor still has to end up correct, and the app has to be told which
     * conversation to reload.
     */
    @Test
    fun `a gap wider than maxAutoRepairSeq is reported instead of backfilled`() = runTest {
        val gateway = gatewayWith()
        val client = newClient(gateway, maxAutoRepairSeq = 3)
        val received = collectMessages(client)
        val reload = collectReloads(client)
        openClient(client, gateway)

        gateway.latest.push(PushTarget.Message, messageJson("c1", 1))
        runCurrent()
        gateway.latest.push(PushTarget.Message, messageJson("c1", 5_000))
        runCurrent()

        assertTrue(gateway.latest.requestsTo("msg.sync").isEmpty(), "4998 messages is not a backfill")
        assertEquals(listOf("c1"), reload)
        assertEquals(listOf(1L, 5_000L), received.map { it.seq })

        // The cursor kept up, so the next message is contiguous rather than another false gap.
        gateway.latest.push(PushTarget.Message, messageJson("c1", 5_001))
        runCurrent()
        assertEquals(listOf(1L, 5_000L, 5_001L), received.map { it.seq })
    }

    // ----------------------------------------------------------------------- resume

    @Test
    fun `reconnecting repairs what conn dot sync reports as gapsFrom`() = runTest {
        val gateway = gatewayWith(
            resumeReply = { request ->
                resumePage(
                    request,
                    conversations = listOf(conversationJson("c1", maxSeq = 8)),
                    gaps = mapOf("c1" to 6L),
                )
            },
            syncReply = { request -> syncPage(request, "c1", listOf(6L, 7L, 8L)) },
        )
        val client = newClient(gateway, store = RecordingCursorStore(ImCursorSnapshot(mapOf("c1" to 5L))))
        val received = collectMessages(client)
        openClient(client, gateway)

        val sync = assertNotNull(gateway.latest.requestsTo("msg.sync").singleOrNull())
        assertEquals(6L, sync.bodyLong("fromSeq"))
        assertEquals(8L, sync.bodyLong("toSeq"))
        assertEquals(listOf(6L, 7L, 8L), received.map { it.seq })
    }

    @Test
    fun `a conversation seen for the first time adopts its position instead of replaying it`() = runTest {
        val gateway = gatewayWith(
            resumeReply = { request ->
                resumePage(request, conversations = listOf(conversationJson("c1", maxSeq = 4_000)))
            },
        )
        val client = newClient(gateway)
        val received = collectMessages(client)
        openClient(client, gateway)

        assertTrue(gateway.latest.requestsTo("msg.sync").isEmpty(), "a cold start is not a gap")
        assertTrue(received.isEmpty())

        // And the adopted cursor is real: the next message is contiguous with it.
        gateway.latest.push(PushTarget.Message, messageJson("c1", 4_001))
        runCurrent()
        assertEquals(listOf(4_001L), received.map { it.seq })
    }

    // --------------------------------------------------------------------- sending

    @Test
    fun `a send without a clientMsgId still gets one`() = runTest {
        val gateway = gatewayWith(
            sendReply = { request ->
                replyTo(
                    request,
                    buildJsonObject {
                        put("messageId", 1)
                        put("seq", 1)
                        put("conversationId", "c1")
                        put("clientMsgId", "echoed")
                        put("createTime", 1)
                    },
                )
            },
        )
        val client = newClient(gateway)
        openClient(client, gateway)

        client.msg.send(SendRequest(target = SendTarget.User("bob"), content = textContent("hello")))
        runCurrent()

        val send = assertNotNull(gateway.latest.requestsTo("msg.send").singleOrNull())
        assertEquals("bob", send.bodyString("receiverId"))
        assertTrue(send.bodyString("clientMsgId").isNotEmpty())
        assertEquals(MessageContentType.Text.code.toLong(), send.bodyLong("contentType"))
    }

    // --------------------------------------------------------------------- helpers

    private fun gatewayWith(
        resumeReply: (JsonObject) -> JsonObject? = { request -> resumePage(request) },
        syncReply: (JsonObject) -> JsonObject? = { null },
        sendReply: (JsonObject) -> JsonObject? = { null },
    ): FakeGateway = FakeGateway().apply {
        onRequest = { request ->
            when (request.target) {
                "conn.sync" -> resumeReply(request)
                "msg.sync" -> syncReply(request)
                "msg.send" -> sendReply(request)
                else -> null
            }
        }
    }
}

// ---- shared harness, used by every ImClient test in this package ----

internal fun TestScope.newClient(
    gateway: FakeGateway,
    maxAutoRepairSeq: Long = 500,
    store: ImCursorStore = RecordingCursorStore(),
    options: ImOptions = testOptions(maxAutoRepairSeq = maxAutoRepairSeq),
): ImClient = ImClient(
    options = options,
    cursorStore = store,
    socketFactory = gateway,
    parentContext = backgroundScope.coroutineContext,
    // Empty rather than Dispatchers.IO: a real dispatcher would escape the test scheduler's
    // virtual time, and every assertion here is about *when* the store was written.
    cursorStoreContext = EmptyCoroutineContext,
)

internal fun TestScope.collectMessages(client: ImClient): List<ImMessage> {
    val received = mutableListOf<ImMessage>()
    backgroundScope.launch { client.messages.collect { received += it } }
    runCurrent()
    return received
}

internal fun TestScope.collectReloads(client: ImClient): List<String> {
    val reloads = mutableListOf<String>()
    backgroundScope.launch { client.conversationNeedsReload.collect { reloads += it } }
    runCurrent()
    return reloads
}

internal fun TestScope.openClient(client: ImClient, gateway: FakeGateway) {
    backgroundScope.launch { client.connect() }
    runCurrent()
    gateway.latest.opened()
    runCurrent()
    kotlin.test.assertEquals(ConnectionState.Open, client.state.value)
}

File: typescript/test/gap-repair.test.ts
import assert from 'node:assert/strict';
import { afterEach, beforeEach, describe, it } from 'node:test';

import { FakeSocket, message, until } from './fake-socket.js';
import { RecordingCursorStore, closeAll, openClient } from './doubles.js';

/**
 * Live gap detection and the resume run, from the application's side.
 *
 * The helpers moved to `doubles.ts` when the cursor store arrived, because five files now need the
 * same "open a client and settle its resume" sequence. The teardown note that used to live here
 * still applies and is worth repeating: an open client holds a 30s heartbeat interval, and in Node
 * a live timer keeps the process alive — leaving one behind makes the run wait out the interval
 * before it can exit.
 */

describe('gap repair', () => {
  beforeEach(() => FakeSocket.reset());
  afterEach(closeAll);

  it('fetches a skipped range before delivering the message that revealed it', async () => {
    const { delivered } = await openClient();
    const socket = FakeSocket.latest;

    socket.push('evt.message', message('s_a_b', 1));
    await until(() => delivered.length === 1);

    // 2 and 3 never arrived.
    socket.push('evt.message', message('s_a_b', 4));
    await until(() => socket.requestsTo('msg.sync').length > 0);

    const repair = socket.requestsTo('msg.sync')[0]!;
    assert.equal(repair.body.conversationId, 's_a_b');
    assert.equal(repair.body.fromSeq, 2);
    assert.equal(repair.body.toSeq, 3);

    socket.reply(repair.id, 'msg.sync', {
      conversationId: 's_a_b',
      messages: [message('s_a_b', 2), message('s_a_b', 3)],
      hasMore: false,
    });

    await until(() => delivered.length === 4);
    // The application never sees the conversation jump forward and then fill in behind it.
    assert.deepEqual(delivered.map((m) => m.seq), [1, 2, 3, 4]);
  });

  it('still delivers the triggering message when the repair itself fails', async () => {
    const { delivered } = await openClient();
    const socket = FakeSocket.latest;

    socket.push('evt.message', message('s_a_b', 1));
    await until(() => delivered.length === 1);

    socket.push('evt.message', message('s_a_b', 4));
    await until(() => socket.requestsTo('msg.sync').length > 0);

    const repair = socket.requestsTo('msg.sync')[0]!;
    socket.reply(repair.id, 'msg.sync', null, 1000, { message: 'boom' });

    // Stalling the conversation because a backfill failed would be worse than the gap. The gap is
    // not lost: `committedSeq` never moved past it, so the next `conn.sync` reports it again.
    await until(() => delivered.length === 2);
    assert.deepEqual(delivered.map((m) => m.seq), [1, 4]);
  });

  it('drops a duplicate seq silently', async () => {
    const { delivered } = await openClient();
    const socket = FakeSocket.latest;

    socket.push('evt.message', message('s_a_b', 1));
    await until(() => delivered.length === 1);

    // Normal after a reconnect replay, and must not reach the application twice.
    socket.push('evt.message', message('s_a_b', 1));
    await new Promise((resolve) => setTimeout(resolve, 60));

    assert.equal(delivered.length, 1);
  });

  it('delivers an unpersisted message (seq 0) without moving the cursor', async () => {
    const { delivered } = await openClient();
    const socket = FakeSocket.latest;

    // Typing, presence and chat-room traffic are never persisted and carry seq 0. Letting one move
    // the cursor would fabricate a gap for the next real message.
    socket.push('evt.message', message('s_a_b', 0));
    socket.push('evt.message', message('s_a_b', 1));

    await until(() => delivered.length === 2);
    assert.deepEqual(delivered.map((m) => m.seq), [0, 1]);
    assert.equal(FakeSocket.latest.requestsTo('msg.sync').length, 0);
  });

  it('accepts the jump and asks for a reload when the gap is larger than the limit', async () => {
    const { delivered, reloads } = await openClient({ maxAutoRepairSeq: 10 });
    const socket = FakeSocket.latest;

    socket.push('evt.message', message('s_a_b', 1));
    await until(() => delivered.length === 1);

    // Replaying fifty thousand messages to catch up helps nobody; the app reloads from history —
    // which it can only do because it is told to. Advancing the cursor in silence, as this SDK
    // used to, leaves a hole nothing will ever mention.
    socket.push('evt.message', message('s_a_b', 50_000));
    await until(() => delivered.length === 2);

    assert.equal(socket.requestsTo('msg.sync').length, 0);
    assert.deepEqual(delivered.map((m) => m.seq), [1, 50_000]);
    assert.deepEqual(reloads, ['s_a_b']);
  });
});

describe('resume after reconnect', () => {
  beforeEach(() => FakeSocket.reset());
  afterEach(closeAll);

  it('repairs every conversation the server reports in gapsFrom', async () => {
    const store = new RecordingCursorStore();
    const { client, delivered } = await openClient({ cursorStore: store });
    const socket = FakeSocket.latest;

    socket.push('evt.message', message('c1', 4));
    await until(() => delivered.length === 1);

    // The application stores it, then says so. Until it does, `convSeqs` reports nothing for this
    // conversation — which is the point of the two cursors, not a gap in the test.
    client.commit('c1', 4);
    await client.flushCursors();

    socket.die();
    await until(() => FakeSocket.instances.length >= 2);

    const next = FakeSocket.latest;
    next.open();
    await until(() => next.requestsTo('conn.sync').length > 0);

    // The client reports where it is, per conversation.
    const sync = next.requestsTo('conn.sync')[0]!;
    assert.deepEqual(sync.body.convSeqs, { c1: 4 });

    // gapsFrom carries only the first seq we are missing; the upper bound is that conversation's
    // maxSeq in the same reply (SPEC-02 §3.4).
    next.reply(sync.id, 'conn.sync', {
      conversations: [{ conversationId: 'c1', maxSeq: 6, updatedAt: 1767225600000 }],
      gapsFrom: { c1: 5 },
      hasMore: false,
    });

    await until(() => next.requestsTo('msg.sync').length > 0);
    const repair = next.requestsTo('msg.sync')[0]!;
    assert.equal(repair.body.fromSeq, 5);
    assert.equal(repair.body.toSeq, 6);

    next.reply(repair.id, 'msg.sync', {
      conversationId: 'c1',
      messages: [message('c1', 5), message('c1', 6)],
      hasMore: false,
    });

    await until(() => delivered.length === 3);
    assert.deepEqual(delivered.map((m) => m.seq), [4, 5, 6]);
  });

  it('adopts the position of a conversation it is seeing for the first time', async () => {
    // Replaying the whole history of a conversation this device has never opened would be the
    // slowest possible way to show a chat list.
    const { client, delivered } = await openClient(
      {},
      {
        conversations: [{ conversationId: 'brand-new', maxSeq: 812, updatedAt: 1767225600000 }],
        gapsFrom: {},
        hasMore: false,
      },
    );

    await new Promise((resolve) => setTimeout(resolve, 40));
    assert.equal(FakeSocket.latest.requestsTo('msg.sync').length, 0);
    assert.deepEqual(delivered, []);

    // And the next real message lands contiguously rather than looking like an 812-wide gap.
    FakeSocket.latest.push('evt.message', message('brand-new', 813));
    await until(() => delivered.length === 1);
    assert.equal(FakeSocket.latest.requestsTo('msg.sync').length, 0);
    assert.equal(client.state, 'open');
  });

  it('leaves cursors untouched when resume fails, so the next message repairs normally', async () => {
    const { client, delivered } = await openClient();
    const socket = FakeSocket.latest;

    socket.push('evt.message', message('c1', 4));
    await until(() => delivered.length === 1);
    client.commit('c1', 4);
    await client.flushCursors();

    socket.die();
    await until(() => FakeSocket.instances.length >= 2);

    const next = FakeSocket.latest;
    next.open();
    await until(() => next.requestsTo('conn.sync').length > 0);
    next.reply(next.requestsTo('conn.sync')[0]!.id, 'conn.sync', null, 1000, { message: 'boom' });

    // Cursor still at 4, so a message at 7 is still recognised as a gap.
    next.push('evt.message', message('c1', 7));
    await until(() => next.requestsTo('msg.sync').length > 0);

    const repair = next.requestsTo('msg.sync')[0]!;
    assert.equal(repair.body.fromSeq, 5);
    assert.equal(repair.body.toSeq, 6);
  });
});

describe('sending', () => {
  beforeEach(() => FakeSocket.reset());
  afterEach(closeAll);

  it('generates a clientMsgId so a naive caller still gets exactly-once semantics', async () => {
    const { client } = await openClient();
    const socket = FakeSocket.latest;

    const pending = client.msg.send({ receiverId: 'bob', content: { text: 'hello' } });
    await until(() => socket.requestsTo('msg.send').length > 0);

    const request = socket.requestsTo('msg.send')[0]!;
    assert.equal(typeof request.body.clientMsgId, 'string');
    assert.ok((request.body.clientMsgId as string).length > 0);
    assert.deepEqual(request.body.content, { text: 'hello' });
    assert.equal(request.body.contentType, 1);

    socket.reply(request.id, 'msg.send', {
      messageId: 9,
      seq: 1,
      conversationId: 's_a_b',
      clientMsgId: request.body.clientMsgId,
      createTime: 1,
    });

    const result = await pending;
    assert.equal(result.seq, 1);
  });

  it('does not redeliver our own message when the server echoes it back', async () => {
    const { client, delivered } = await openClient();
    const socket = FakeSocket.latest;

    const pending = client.sendText({ receiverId: 'bob' }, 'hello');
    await until(() => socket.requestsTo('msg.send').length > 0);

    const request = socket.requestsTo('msg.send')[0]!;
    socket.reply(request.id, 'msg.send', {
      messageId: 9,
      seq: 1,
      conversationId: 's_a_b',
      clientMsgId: request.body.clientMsgId,
      createTime: 1,
    });
    await pending;

    // Sending advanced deliveredSeq to 1, so the echo is a duplicate and is dropped. It did *not*
    // advance committedSeq: the application has not stored it yet, and only it knows when it has.
    socket.push('evt.message', message('s_a_b', 1));
    await new Promise((resolve) => setTimeout(resolve, 60));

    assert.deepEqual(delivered, []);
    assert.equal(client.deliveredSeq('s_a_b'), 1);
    assert.equal(client.committedSeq('s_a_b'), 0);
  });
});
